### DISCUSSION TITLE
`gltf-transform optimize` removes `uv1` attribute from geometry.

### DISCUSSION THREAD

**Discussion by subhankar-trisetra** [donmccurdy/glTF-Transform#1296] [Q&A] [answered]
Here are the differences in the geometry attributes before and after optimization using gltf-transform.
Before:
<img width="607" alt="before_compression" src="https://github.com/donmccurdy/glTF-Transform/assets/98544661/6c9629a7-4018-4df8-899d-5ac69b14d101">
After:
**Notice the missing `uv1` property form here**
![with_texture_compression](https://github.com/donmccurdy/glTF-Transform/assets/98544661/a7fba06d-125f-4525-b70a-68fb73eb882f)

Command I used:
```bash
gltf-transform optimize without_compression.glb scene_opt.glb --compress draco --flatten false --join false --palette false --texture-compress auto
```

Is there a way to fix this?

**UPDATE**:

If we run each step separately, it seems to preserve the `uv1` channel and the final output is 1mb greater than just running the optimize command e.g.:

```bash
gltf-transform dedup without_compression.glb scene_opt.glb
gltf-transform instance scene_opt.glb scene_opt.glb
gltf-transform weld scene_opt.glb scene_opt.glb
gltf-transform simplify scene_opt.glb scene_opt.glb
gltf-transform resample scene_opt.glb scene_opt.glb
gltf-transform prune scene_opt.glb scene_opt.glb
gltf-transform sparse scene_opt.glb scene_opt.glb
gltf-transform png scene_opt.glb scene_opt.glb
gltf-transform jpeg scene_opt.glb scene_opt.glb
gltf-transform draco scene_opt.glb scene_opt.glb
```

The issue that I'm majorly facing is: `uv1` is essential for me to store the lightmap data, and necessary for lightmaps to get applied in threejs. uv0 channel is kept for the uvs. 

**UPDATE**

Upon doing some digging, what I found is: https://github.com/donmccurdy/glTF-Transform/blob/255b71561720817e797a165c4be8ef88a0794d0b/packages/cli/src/cli.ts#L361-L366

And I believe the `keepAttributes: false` property in prune function is responsible for removing the `uv1` attributes. Correct me if I'm wrong over here.

**Comment by donmccurdy** [comment, marked as answer]
Hi @subhankar-trisetra! You're correct that the `keepAttributes: false` option is responsible for removing UVs (and other vertex attributes) that are unused in the given file. Currently the `optimize` CLI command doesn't expose that as a flag. If you're doing this programmatically in a script you can of course choose whichever options, or I'd be open to PRs adding this flag to the `optmize` CLI command.

**Reply by subhankar-trisetra** [reply]
Hi @donmccurdy  As I'm putting up the PR, just wanted to know what would be the correct way to accept the params for the prune function, using a stringified JSON as an argument or separate arguments for each params.
i.e. `--prune-options '{"keepAttributes":false}'` or `--prune-attributes true --prune-indices false`

**Reply by subhankar-trisetra** [reply]
Here's my PR: https://github.com/donmccurdy/glTF-Transform/pull/1298

I went with option 2 for the above query. Please do let me know if anything needs to be changed.
